### PR TITLE
dotnet-sdk_7: 7.0.102 -> 7.0.201

### PR DESCRIPTION
**Pull request by raphaelr**
###### Description of changes

Release notes: https://github.com/dotnet/core/blob/main/release-notes/7.0/7.0.3/7.0.3.md, https://github.com/dotnet/core/blob/main/release-notes/7.0/7.0.3/7.0.201.md

Fixes: CVE-2023-21808

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by mdarocha — APPROVED** [review]

**Comment by JamieMagee** [comment]
Result of `nixpkgs-review pr 219205` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
    <li>BeatSaberModManager</li>
    <li>denaro</li>
    <li>dotnet-aspnetcore_7 (dotnetCorePackages.aspnetcore_7_0)</li>
    <li>dotnet-runtime_7 (dotnetCorePackages.runtime_7_0)</li>
    <li>dotnet-sdk_7 (dotnetCorePackages.sdk_7_0)</li>
    <li>pinta</li>
    <li>ryujinx</li>
    <li>wasabibackend</li>
  </ul>
</details>

**Review by GabrielDougherty — APPROVED** [review]

**Comment by GabrielDougherty** [comment]
Result of `nixpkgs-review pr 219205` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
    <li>BeatSaberModManager</li>
    <li>denaro</li>
    <li>dotnet-aspnetcore_7 (dotnetCorePackages.aspnetcore_7_0)</li>
    <li>dotnet-runtime_7 (dotnetCorePackages.runtime_7_0)</li>
    <li>dotnet-sdk_7 (dotnetCorePackages.sdk_7_0)</li>
    <li>pinta</li>
    <li>ryujinx</li>
    <li>wasabibackend</li>
  </ul>
</details>

File: pkgs/games/BeatSaberModManager/deps.nix [hunk 1]
```diff
@@ -26,21 +26,21 @@
   (fetchNuGet { pname = "HarfBuzzSharp.NativeAssets.Win32"; version = "2.8.2.1-preview.108"; sha256 = "0n6ymn9jqms3mk5hg0ar4y9jmh96myl6q0jimn7ahb1a8viq55k1"; })
   (fetchNuGet { pname = "JetBrains.Annotations"; version = "10.3.0"; sha256 = "1grdx28ga9fp4hwwpwv354rizm8anfq4lp045q4ss41gvhggr3z8"; })
   (fetchNuGet { pname = "MicroCom.Runtime"; version = "0.11.0"; sha256 = "0p9c3m0zk59x9dcqw077hzd2yk60myisbacvm36mnwpcjwzjkp2m"; })
-  (fetchNuGet { pname = "Microsoft.AspNetCore.App.Ref"; version = "6.0.13"; sha256 = "1k49nrz5rw1yglkjypmmnclahib2qiwm838008yahk00ysy039f4"; })
-  (fetchNuGet { pname = "Microsoft.AspNetCore.App.Runtime.linux-arm64"; version = "6.0.13"; sha256 = "1a80w34yb700cvjzfcsq3c5sb32bk95ccgjfcvyqxkpzlgcc50xn"; })
-  (fetchNuGet { pname = "Microsoft.AspNetCore.App.Runtime.linux-x64"; version = "6.0.13"; sha256 = "1y2gnkrc788y9k79sygx7dmgkj674qwdajyrxizx0a70rwmjszmw"; })
+  (fetchNuGet { pname = "Microsoft.AspNetCore.App.Ref"; version = "6.0.14"; sha256 = "0q43lxc5wdw5vaypzc068yx8q1s85sj3yw1lcdjr0ps7nzzv4laa"; })
```
Comment (LeSuisse) [inline pkgs/games/BeatSaberModManager/deps.nix:29]: Are those changes related to the 7.0.3 upgrade?

**Review by LeSuisse — CHANGES_REQUESTED** [review]

File: pkgs/games/BeatSaberModManager/deps.nix (hunk 1 above)
Comment (raphaelr) [reply pkgs/games/BeatSaberModManager/deps.nix:29]: Unfortunately yes, BeatSaberModManager doesn't build against the new dotnet version unless I also update BeatSaberModManager's nuget dependencies.

I've already split the changes into two commits.

Comment (SuperSandro2000) [reply pkgs/games/BeatSaberModManager/deps.nix:29]: thats normal for dotnet

**Review by SuperSandro2000 — APPROVED** [review]

**Review by LeSuisse — APPROVED** [review]

**Comment by Atemu** [comment]
Thanks for bumping BSMM deps btw. I happened to see this PR and made a TODO to check BSMM but you already took care of that :)

Feel free to ping me next time.

**Comment by mdarocha** [comment]
> Thanks for bumping BSMM deps btw. I happened to see this PR merged and made a TODO to check BSMM but you already took care of that :)
> 
> Feel free to ping me next time.

Do you know why this package requires manual dependency updates after dotnet update? The depedencies updated should be included implicitly by `buildDotnetModule`.

**Comment by Atemu** [comment]
No idea whatsoever. I don't have any DotNet experience. I just want to play Beat Saber with mods ;)

**Comment by raphaelr** [comment]
> The depedencies updated should be included implicitly by buildDotnetModule.

Yes, but BeatSaberModManager is a .NET 7 project so we build it with dotnet-sdk_7. However, it depends on a .NET 6 library ([BSIPA-Linux](https://github.com/geefr/BSIPA-Linux/blob/81788f58b75f0510a2c534ebe34f11827e167e07/IPA-lib/IPA-lib.csproj)).

Because of this BSMM depends on some packages twice... For example, it depends on `Microsoft.NETCore.App.Ref` 6.0.14 *and* 7.0.3.

This problem should be solvable by improving buildDotnetModule. In theory setting `dotnet-sdk = with dotnetCorePackages; combinePackages [ sdk_6_0 sdk_7_0 ]` *should* work and do the right thing, but it currently doesn't because the thing returned by dotnetCorePackages.combinePackages (a wrapper around `buildEnv`) lacks the neccessary metadata for buildDotnetModule to work...

An alternative would be to patch [this line](https://github.com/geefr/BSIPA-Linux/blob/81788f58b75f0510a2c534ebe34f11827e167e07/IPA-lib/IPA-lib.csproj#L5) (from net6.0 to net7.0) in patchPhase, but it's a bit unsafe because there are breaking changes between major .NET versions.

**Comment by mdarocha** [comment]
> > The depedencies updated should be included implicitly by buildDotnetModule.
> 
> Yes, but BeatSaberModManager is a .NET 7 project so we build it with dotnet-sdk_7. However, it depends on a .NET 6 library ([BSIPA-Linux](https://github.com/geefr/BSIPA-Linux/blob/81788f58b75f0510a2c534ebe34f11827e167e07/IPA-lib/IPA-lib.csproj)).
> 
> Because of this BSMM depends on some packages twice... For example, it depends on `Microsoft.NETCore.App.Ref` 6.0.14 _and_ 7.0.3.
> 
> This problem should be solvable by improving buildDotnetModule. In theory setting `dotnet-sdk = with dotnetCorePackages; combinePackages [ sdk_6_0 sdk_7_0 ]` _should_ work and do the right thing, but it currently doesn't because the thing returned by dotnetCorePackages.combinePackages (a wrapper around `buildEnv`) lacks the neccessary metadata for buildDotnetModule to work...
> 
> An alternative would be to patch [this line](https://github.com/geefr/BSIPA-Linux/blob/81788f58b75f0510a2c534ebe34f11827e167e07/IPA-lib/IPA-lib.csproj#L5) (from net6.0 to net7.0) in patchPhase, but it's a bit unsafe because there are breaking changes between major .NET versions.
> 
> Or patch the nuget-to-nix script that generates deps.nix to ignore the version when checking if it should download a package. So that it wouldn't attempt to download version 6 of a package if and only if the same version or higher is already part of dotnet-sdk.

Thanks for the info, I'll attempt a fix for that.

**Pull request merged**